### PR TITLE
Enabled `full_output` and using reference PIE to validate our PIEs are correct

### PR DESCRIPTION
**Pull request by shamsasari**
The reference PIE was created with `full_output` enabled, so `output.rs` was updated to support that. The code was also refactored to better mirror the Cairo serialization code.

### CODE REVIEW

File: scripts/prove-blocks.sh
[the remark this answers is not in the document]
Comment (shamsasari) [reply scripts/prove-blocks.sh:1]: It's not used as far as I can see. It's not referenced anywhere in the code base, and seems to be a leftover when prove_blocks.yml was introduced.

[the remark this answers is not in the document]
Comment (shamsasari) [reply scripts/prove-blocks.sh:1]: OK, I will revert.

File: .github/workflows/prove_blocks.yml [hunk 1]
```diff
@@ -50,7 +50,7 @@ jobs:
         env:
           PATHFINDER_RPC_URL: ${{ secrets.PATHFINDER_RPC_URL }}
         run: |
-          cargo test --release --package prove_block --test prove_block -- test_prove_selected_blocks --show-output --ignored
+          cargo test --release --package prove_block --test prove_block -- --show-output --ignored
```
[the remark this answers is not in the document]
Comment (shamsasari) [reply .github/workflows/prove_blocks.yml:53]: Actually, I don't like how these ignored tests are used in the first place.

[the remark this answers is not in the document]
Comment (shamsasari) [reply .github/workflows/prove_blocks.yml:53]: Then just make the test conditional on the presence of `PATHFINDER_RPC_URL`? Seems you can do that with [feature flags](https://stackoverflow.com/questions/73760692/ignore-a-test-based-on-environment-variable).

[the remark this answers is not in the document]
Comment (shamsasari) [reply .github/workflows/prove_blocks.yml:53]: No, for this exact line we're commenting on! It can just be 
```
cargo test --release --package prove_block --test prove_block -- --show-output
```
and it will run the tests in release mode for just the `prove_block` tests.

[the remark this answers is not in the document]
Comment (shamsasari) [reply .github/workflows/prove_blocks.yml:53]: It's not a change for this PR anyway.

File: crates/bin/prove_block/src/main.rs [hunk 2]
```diff
@@ -24,15 +22,15 @@ fn init_logging() {
...
 
     let args = Args::parse();
 
     let block_number = args.block_number;
     let layout = LayoutName::all_cairo;
 
-    let result = prove_block::prove_block(block_number, &args.rpc_provider, layout).await;
-    let (_pie, _output) = result.map_err(debug_prove_error).expect("Block could not be proven");
-    Ok(())
+    let result = prove_block::prove_block(block_number, &args.rpc_provider, layout, false).await;
```
Comment (HermanObst) [inline crates/bin/prove_block/src/main.rs:33]: @shamsasari why are you setting the full output as false?
It makes sense to have it in true as this is the default use case Starkware need for the SNOS.

File: crates/starknet-os/src/io/output.rs [hunk 3]
```diff
@@ -134,99 +134,140 @@ fn next_as_usize<I: Iterator<Item = Felt252>>(output_iter: &mut I, item_name: &s
...
+pub fn deserialize_os_output<I>(output_iter: &mut I) -> Result<StarknetOsOutput, SnOsError>
+where
+    I: Iterator<Item = Felt252>,
+{
+    let header = read_segment(output_iter, HEADER_SIZE, "header elements")?;
+    let use_kzg_da = header[USE_KZG_DA_OFFSET];
+    let full_output = header[FULL_OUTPUT_OFFSET];
 
-    if use_kzg_da {
+    if !use_kzg_da.is_zero() {
```
Comment (HermanObst) [inline crates/starknet-os/src/io/output.rs:270]: What is the difference between this and just `use_kzg_da`?

File: crates/starknet-os/src/io/output.rs [hunk 4]
```diff
@@ -239,13 +280,13 @@ pub fn decode_output<I: Iterator<Item = Felt252>>(output_iter: &mut I) -> Result
         let _: Vec<_> = output_iter.by_ref().take(2 * 2 * n_blobs).collect();
```
Comment (HermanObst) [inline crates/starknet-os/src/io/output.rs:280]: Why are we skipping this data? 
@shamsasari 

cc: @notlesh

**Review by HermanObst — CHANGES_REQUESTED** [review]
@shamsasari looking good! Super good job here :) 

I left some comments to discuss, and I would also need you to add an unit test with `full_output: Felt252::ZERO`

File: crates/starknet-os/src/io/output.rs (hunk 3 above)
Comment (shamsasari) [reply crates/starknet-os/src/io/output.rs:270]: `use_kzg_da` used to be `bool` converted from the Felt. I've kept the original Felt value and done a zero check, mirroring what the Cairo code is doing.

**Comment by HermanObst** [comment]
@shamsasari also, I would love some extensive documentation on:
- deserialize_contract_state_inner
- deserialize_contract_class_da_changes
- deserialize_contract_state
- deserialize_da_changes

File: crates/bin/prove_block/src/main.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (shamsasari) [reply crates/bin/prove_block/src/main.rs:33]: Even though it sets `use_kzg` to false? I thought KZG was necessary.

**Comment by shamsasari** [comment]
> I would also need you to add an unit test with `full_output: Felt252::ZERO` in crates/starknet-os/src/io/output.rs

The unit test is already using `full_output: Felt252::ZERO`, or did you mean 1? Also, the unit test is not testing the deserialization code at all, just the JSON format of the `StarknetOsOutput`.

**Review by HermanObst — APPROVED** [review]

**Comment by ftheirs** [comment]
Looks great! I just have a similar comment than Herman said. If you could add a link to the Cairo files that you're mirroring would be great!

File: crates/starknet-os/src/io/output.rs [hunk 5]
```diff
@@ -134,99 +134,141 @@ fn next_as_usize<I: Iterator<Item = Felt252>>(output_iter: &mut I, item_name: &s
...
-        Felt252::from(1u128 << 64).try_into().expect("2**64 should be considered non-zero. Did you change the value?");
-    let (class_nonce, n_changes) = class_nonce_n_changes.div_rem(&two_exp_64);
-    let (class_updated, nonce) = class_nonce.div_rem(&two_exp_64);
+    let value = next_or_fail(output_iter, "contract change value")?;
+    let (value, n_actual_updates) = value.div_rem(&bound);
+    let (was_class_updated, new_state_nonce) = value.div_rem(&bound);
 
-    let class_hash = if class_updated != Felt252::ZERO {
-        Some(next_or_fail(output_iter, "contract change class_hash")?)
+    #[allow(clippy::collapsible_else_if)] // Mirror the Cairo code as much as possible
```
Comment (ftheirs) [inline crates/starknet-os/src/io/output.rs:151]: Probably here is a good point to add the link to cairo-lang file. This would be a good reference if there are changes in future versions from cairo-lang

**Review by ftheirs — CHANGES_REQUESTED** [review]

File: crates/starknet-os/src/io/output.rs (hunk 5 above)
Comment (shamsasari) [reply crates/starknet-os/src/io/output.rs:151]: Yes I've done that at the top of this function.

**Comment by shamsasari** [comment]
I've linked all remaining functions

**Review by ftheirs — APPROVED** [review]
